### PR TITLE
Improve cookies

### LINKED ISSUE #2969: Cookies set do not use the secure flag
When using HTTPS all cookies set by Etherpad should get the [secure flag](https://www.owasp.org/index.php/SecureFlag), which prevents that this cookies are leaked in insecure (HTTP) requests.

I tried adding `proxy_set_header X-Forwarded-Proto $scheme;` to my nginx config, but NodeJS still does not add the secure flag.
So I think the easiest way to solve this would be to add a setting to the `settings.json`:  `"cookiesAddSecureFlag": true;`

This way web admins can configure this manually.

The same also applies to the [`HttpOnly`](https://www.owasp.org/index.php/HttpOnly) flag, ~~but AFAIK you need to read all cookies via JavaScript, don't you?~~ Okay, for `express_sid` you already do this. What about other cookies such as `token`? (For what is this cookie used anyway?)
And if I'm already complaining about these cookies: Why does Etherpad set a cookie `test` with the value `test`?

### PR DESCRIPTION
**Pull request by Gared**
Fixes #2969 
- [x] Remove "test" cookie
- [x] Add secure flag to browser created cookies if pad accessed through https
- [x] Make express-cookie scheme dependent

### CODE REVIEW

**Comment by rugk** [comment]
Also partially fixes https://github.com/ether/etherpad-lite/issues/2985, does not it? At least the test cookie get's away. Taht's good! :smiley: 
Still the path value of `token` is set incorrectly and alos this `debug` in the local storage is unnecessary.

**Comment by Gared** [comment]
@rugk Yeah, this fixes this ticket partially. One step after another :-)

**Comment by Gared** [comment]
@rugk Can you please test this ticket with your reverse proxy?

File: src/static/js/pad_cookie.js [hunk 1]
```diff
@@ -79,6 +80,10 @@ var padcookie = (function()
       alreadyWarnedAboutNoCookies = true;
     }
   }
+  
+  function isHttpsScheme() {
+    return window.location.protocol == "https:";
```
Comment (rugk) [inline src/static/js/pad_cookie.js]: I think this won't work for reverse proxys as they usually forward requests over HTTP. You should honor the [`X-Forwarded-Proto`](https://stackoverflow.com/questions/13111080/what-is-a-full-specification-of-x-forwarded-proto-) header - and probably also the standardized [`Forwarded`](https://tools.ietf.org/html/rfc7239) header, but AFAIK the unofficial `X-Forwarded-Proto` is still more common.

File: src/static/js/pad_utils.js [hunk 2]
```diff
@@ -53,13 +53,16 @@ function createCookie(name, value, days, path){ /* Used by IE */
   if(!path){ // IF the Path of the cookie isn't set then just create it on root
     path = "/";
   }
+  
+  //Check if we accessed the pad over https
+  var secure = window.location.protocol == "https:" ? ";secure" : "";
```
Comment (rugk) [inline src/static/js/pad_utils.js]: Same as above. This does not honor reverse proxys.

**Comment by rugk** [comment]
I can do that soon. However I already commented on some source code lines where I see that this won't work, so could you please fix this first?

**Comment by rugk** [comment]
For now I already [added this header](https://github.com/ether/etherpad-lite/wiki/How-to-put-Etherpad-Lite-behind-a-reverse-Proxy/_compare/bdb85b0188370354676fa47ca406fbabd9a588e8...16d5a32c0707ade0cbbf5d1b3045793949b1b906) to the config.

**Comment by Gared** [comment]
@rugk Thanks for adjusting the config. The comments in my code are the part which is executed directly in the browser. So the browser knows if it is talking http or https with the server. The cookie which is set on server-side is for express.

**Comment by rugk** [comment]
Ah okay, so I'll test it soon.

**Comment by rugk** [comment]
At first I had some issues with caching: You should consider loading the files with `file.css?v=0.3`.

So but now when I see it working:
* test is gone.
* SecureFlag is set for `prefs` and `token`, but **not** for `express_sid`, `io` and `lang`.

Additionally I get a "cookies could not be found" in the browser when Etherpad is loading (that's before the cookies with SecureFlag are set), but it disappears quickly. Anyway it is a message, which is at least confusing.

And `express_sid`is still the only cookie with HTTPOnly flag.

**Comment by DanielBickley** [comment]
@rugk @Gared 

You're getting the "cookies could not be found message" because the cookie test is failing on "prefs" basically because it doesn't exist at that point. You could conduct the test after the "prefs" cookie is actually attempted to be created, or you could do what I do and just create the test cookie, and if found to exist, then promptly delete the cookie.

**Comment by rugk** [comment]
Any news about this PR? @Gared could @DanielBickley's advice help you?

**Pull request merged**